File: 5KYU/Alphabet war - Wo lo loooooo priests join the war/solution.cs
using System.Collections.Generic;
using System.Linq;
using System;

public class Kata
{
    static Dictionary<char, int> leftScores = new()
    {
        {'w', 4},{'p', 3},{'b', 2},{'s', 1}
    };

    static Dictionary<char, int> rightScores = new()
    {
        {'m', 4},{'q', 3},{'d', 2},{'z', 1}
    };

    static char ConvertLetter(char c) =>
        c switch
        {
            'w' => 'm',
            'm' => 'w',
            'q' => 'p',
            'p' => 'q',
            'b' => 'd',
            'd' => 'b',
            'z' => 's',
            's' => 'z',
             _ => c
        };

     static bool CanConvertLeft(char[] xs, int i, char priest)
     {
         bool AdjacentIsFriendly() =>
             (priest == 't' && "wpbs".Contains(xs[i-1])) || 
             (priest == 'j' && "mqdz".Contains(xs[i-1]));

         bool AdjacentSurroundedByEnemyPriest() =>
             (priest == 'j' && xs[i - 2] == 't') ||
             (priest == 't' && xs[i - 2] == 'j');

         if (i == 0)
            return false;

         if (i == 1 && !AdjacentIsFriendly())
            return true;

         return !AdjacentIsFriendly() && !AdjacentSurroundedByEnemyPriest();
     }

     static bool CanConvertRight(char[] xs, int i, char priest)
     {
         bool AdjacentIsFriendly() =>
            (priest == 't' && "wpbs".Contains(xs[i+1])) || 
            (priest == 'j' && "mqdz".Contains(xs[i+1]));

         bool AdjacentSurroundedByEnemyPriest() =>
            (priest == 'j' && xs[i + 2] == 't') ||
            (priest == 't' && xs[i + 2] == 'j');
    
         if (i == xs.Length-1)
            return false;

         if (i == xs.Length-2 && !AdjacentIsFriendly())
            return true;

         return !AdjacentIsFriendly() && !AdjacentSurroundedByEnemyPriest();
     }

     static string ConvertPriests(string fight)
     {
        var xs = fight.ToCharArray();
        var res = fight.ToCharArray();
        for (var i = 0; i < xs.Length; i++)
        {
            // COnvert left
            if ((xs[i] == 'j' || xs[i] == 't') && i > 0 && CanConvertLeft(xs, i, xs[i]))
               res[i - 1] = ConvertLetter(xs[i - 1]);

            if ((xs[i] == 'j' || xs[i] == 't') && i < xs.Length - 1 && CanConvertRight(xs, i, xs[i]))
               res[i + 1] = ConvertLetter(xs[i + 1]);
        }
    
        return string.Join("", res);
      }
   
      public static string AlphabetWar(string fight)
      {
          fight = ConvertPriests(fight);
        
          var leftScore = 0;
          var rightScore = 0;

          foreach (var x in fight.ToCharArray())
          {
               if (leftScores.ContainsKey(x))
                  leftScore += leftScores[x];
              if (rightScores.ContainsKey(x))
                  rightScore += rightScores[x];
          }
      
          if(leftScore > rightScore)
              return "Left side wins!";
          if(leftScore < rightScore)
              return "Right side wins!";
      
          return "Let's fight again!";
      }
 }

File: 5KYU/Alphabet war - Wo lo loooooo priests join the war/solution.go
package kata

var LeftPriest = 't'
var RightPriest = 'j'

var LeftPowers = map[rune]int{
	'w': 4,
	'p': 3,
	'b': 2,
	's': 1}

var RightPowers = map[rune]int{
	'm': 4,
	'q': 3,
	'd': 2,
	'z': 1}

var opposites = map[rune]rune{
	'w': 'm',
	'p': 'q',
	'b': 'd',
	's': 'z',
	'm': 'w',
	'q': 'p',
	'd': 'b',
	'z': 's'}

func Score(fight string) (int, int) {
	left, right := 0, 0

	for _, c := range fight {
		if _, ok := LeftPowers[c]; ok {
			left += LeftPowers[c]
		}

		if _, ok := RightPowers[c]; ok {
			right += RightPowers[c]
		}
	}
	return left, right
}

func GetNeighbours(fight string, idx int) (rune, rune) {
	leftNeighbour, rightNeighbour := ' ', ' '
	length := len(fight)
	if idx > 0 {
		leftNeighbour = rune(fight[idx-1])
	}

	if idx < length-1 {
		rightNeighbour = rune(fight[idx+1])
	}

	return leftNeighbour, rightNeighbour
}

func CanConvert(c rune, leftNeighbour rune, rightNeighbour rune) bool {
	if _, ok := LeftPowers[c]; ok {
		if (leftNeighbour == RightPriest && rightNeighbour != LeftPriest) ||
			(rightNeighbour == RightPriest && leftNeighbour != LeftPriest) {
			return true
		}
	} else if _, ok := RightPowers[c]; ok {
		if (leftNeighbour == LeftPriest && rightNeighbour != RightPriest) ||
			(rightNeighbour == LeftPriest && leftNeighbour != RightPriest) {
			return true
		}
	}
	return false
}

func AlphabetWar(fight string) string {
	res := ""
	for idx, c := range fight {
		leftNeighbour, rightNeighbour := GetNeighbours(fight, idx)
		if CanConvert(c, leftNeighbour, rightNeighbour) {
			res += string(opposites[c])
		} else {
			res += string(c)
		}
	}

	leftScore, rightScore := Score(res)
	if leftScore > rightScore {
		return "Left side wins!"
	} else if rightScore > leftScore {
		return "Right side wins!"
	} else {
		return "Let's fight again!"
	}
}
